### ISSUE TITLE
Detecting disjoint sets in conditional types

### ISSUE THREAD

**Issue by krryan** [Microsoft/TypeScript#23714]
**TypeScript Version:**  2.7.0-dev.20180412

**Search Terms:** disjoint, exclusive, intersection

**Code**

```ts
type UnionOfDisjointStrings<One extends string, Another extends string> =
    One & Another extends never
        ? One | Another
        : never;

type AB = UnionOfDisjointStrings<'A', 'B'>; // 'A' | 'B' as desired
type CC = UnionOfDisjointStrings<'C', 'C'>; // never as desired
type ACC = UnionOfDisjointStrings<'A' | 'C', 'C'>; // 'A' | 'B' !NOT desired!
type ABAC = UnionOfDisjointStrings<'A' | 'B', 'A' | 'C'>; // 'A' | 'B' | 'C' !NOT desired!
```

**Desired behavior:**
`AB` has type `'A' | 'B'`, `CC`, `ACC`, and `ABAC` all have type `never`.

**Actual behavior:**
`AB` and `CC` are correctly `'A' | 'B'` and `never`, respectively

`ACC` is `'A' | 'C'` and `ABAC` is `'A' | 'B' | 'C'`.

**Playground Link:** All four types evaluate to `never` on the playground.

**Related Issues:** #14094

**Discussion:**
Note that "desired" is used here rather than "expected" because I understand _why_ things are this way and why these results are as expected. In short, the distribution of conditional types is the source of the problems. I think the type transformations progress as follows:
```ts
UnionOfDisjointStrings<'A' | 'B', 'A' | 'C'>
```
```ts
UnionOfDisjointStrings<'A', 'A'> | UnionOfDisjointStrings<'A', 'C'> | UnionOfDisjointStrings<'B', 'A'> | UnionOfDisjointStrings<'B', 'C'>
```
```ts
never | 'A' | 'C' | 'B' | 'A' | 'B' | 'C'
```
```
'A' | 'C' | 'B'
```

But as a consequence, there does not seem to be any way to detect if two unions of string literals are disjoint (that is, no members of one appear in the other). Using `One extends Another ? never : Another extends One ? never : One | Another` works if one of them is strictly a subset of the other, but not if the two unions share some elements while having other unique elements.

There really ought to be some way of accomplishing this. My current use-case has to do with `RegExp` (specifically, compile-time checks that you don't re-use capture groups), but this is a pretty basic operation...

**Comment by mhegazy** [comment]
Well.. all of them are `never` on 2.8.3 and and 2.9, ([playground is running 2.8](http://www.typescriptlang.org/play/index.html#src=type%20UnionOfDisjointStrings%3COne%20extends%20string%2C%20Another%20extends%20string%3E%20%3D%0D%0A%20%20%20%20One%20%26%20Another%20extends%20never%0D%0A%20%20%20%20%3F%20One%20%7C%20Another%0D%0A%20%20%20%20%3A%20never%3B%0D%0A%0D%0Atype%20AB%20%3D%20UnionOfDisjointStrings%3C'A'%2C%20'B'%3E%3B%20%2F%2F%20'A'%20%7C%20'B'%20as%20expected%0D%0Atype%20CC%20%3D%20UnionOfDisjointStrings%3C'C'%2C%20'C'%3E%3B%20%2F%2F%20never%20as%20expected%0D%0Atype%20ACC%20%3D%20UnionOfDisjointStrings%3C'A'%20%7C%20'C'%2C%20'C'%3E%3B%20%2F%2F%20'A'%20%7C%20'B'%20!NOT%20expected!%0D%0Atype%20ABAC%20%3D%20UnionOfDisjointStrings%3C'A'%20%7C%20'B'%2C%20'A'%20%7C%20'C'%3E%3B%20%2F%2F%20'A'%20%7C%20'B'%20%7C%20'C'%20!NOT%20expected)).

intersection types are not reduced until they are added to a union..

This might do what you are looking for:

```ts
type UnionOfDisjointStrings<One extends string, Another extends string> = [One, Another] extends [Exclude<One, Another>, Exclude<Another, One>] ? One | Another : never;

```

**Comment by krryan** [comment]
Using tuples is an interesting idea! I will explore that, though I worry about order mattering then.

Anyway, I've updated my 'bug' and made it more of a 'suggestion'. It took some time but I worked out how and why things happened this way—it's a necessary consequence of distributing conditional types. Which is obviously desirable in other situations. But nonetheless, there ought to be some way to express this kind of thing.
